Fix #264: New Rule: Private Data Types (#300)
inaka/elvis_core: doc_rules/elvis_style/private_data_types.md

# Private Data Types

(since [3.0.0](https://github.com/inaka/elvis_core/releases/tag/3.0.0))

> Exporting a function's internal data type in order to construct those types

Reviewer: ```suggestion
Exporting a function's internal data types in order to construct those types
```

> externally results in tightly-coupled code. A module's should be responsible

Reviewer: ```suggestion
externally results in tightly-coupled code. A module should be responsible
```

> Exporting a function's internal data type in order to construct those types
> externally results in tightly-coupled code. A module's should be responsible
> for creating its own internal types. So, if the data type is needed outside the
> module, it should be
> [an opaque type](https://www.erlang.org/doc/reference_manual/opaques.html).

Reviewer: ```suggestion
Exporting functions' internally-defined data types, in order to consume those
types externally, results in tightly-coupled code. Modules should be responsible
for defining their own internal data types. If these are needed outside the
modules, they should be made
[opaque](https://www.erlang.org/doc/reference_manual/opaques.html).
```

> "Works on `.beam` file? Yes!"

## Options

> - `applies_to :: [atom()]`

Reviewer 2: ```suggestion
- `applies_to :: [record | map | tuple]`
```

> - `applies_to :: [atom()]`

Reviewer: I'd rename the option `apply_to`. Thoughts?

Reviewer 2: I'm fine with `applies_to`. I read it as "The rule applies to…".

I don't mind `apply_to` (as in "apply this rule to…"), but if it's a matter of taste, I would keep "applies".

Reviewer: I'd say it's a question of consistency:

https://github.com/inaka/elvis_core/blob/main/doc_rules/elvis_style/max_function_length.md 
- count, count

https://github.com/inaka/elvis_core/blob/main/doc_rules/elvis_style/no_macros.md
- allow

https://github.com/inaka/elvis_core/blob/main/doc_rules/elvis_text_style/no_trailing_whitespace.md
- ignore

Reviewer 2: Good point.

Reviewer: cf768b4

  - default: `record`.

>   - other options: `map` and `tuple`.

Reviewer: ```suggestion
```

Reviewer: Not required, since we're already mentioning all options above.
